### PR TITLE
[srp] implement `AdvertisingProxy` and define `Dnssd` platform APIs

### PR DESCRIPTION
**Pull request by abtink**
This commit adds a generic SRP Advertising Proxy implementation to OpenThread core, which uses a set of newly defined `otPlatDnssd` platform APIs for DNS-SD (mDNS) support on infrastructure network on a Border Router.

`Srp::Server` provides `ServiceUpdateHandler` callback mechanism that allows platforms to implement their own advertising proxy function. While this is still supported, the new generic advertising proxy implementation makes it easier to port and support the proxy function on new platforms. The platform needs to provide the DNS-SD platform APIs, which are designed to be simple and easy to implement.

The `AdvertisingProxy` directly interacts with `Srp::Server` and its registered `Host` and `Service` entries, tracking whether an entry has been successfully advertised, is currently being advertised, or has been replaced by a new registration.

The `AdvertisingProxy` ensures that consecutive SRP updates for the same host or service are committed on the server in the order they are received, even if their advertisements are finished in a different order. This is important for SRP Replication support, as the server may receive a large number of SRP updates back-to-back for the same host.

The `AdvertisingProxy` will also register key records for SRP host and service instance names. This will keep the claim on the name of a removed entry while its key lease is not expired. It is also used when an SRP host registration has no off-mesh routable address.
    
This commit adds a detailed unit test `test_srp_adv_proxy` that validates the `AdvertisingProxy` under many scenarios. The test covers a range of cases, including delayed registration callbacks and timeouts, new registrations replacing outstanding advertisements, platform DNS-SD state changes and failures, host address changes adding/removing OMR addresses.

----

~This PR currently contains the commit from https://github.com/openthread/openthread/pull/9208.~

### CODE REVIEW

**Comment by wgtdkp** [comment]
> This commit adds a detailed unit test test_srp_adv_proxy that validates the AdvertisingProxy under many scenarios. The test covers a range of cases, including delayed registration callbacks and timeouts, new registrations replacing outstanding advertisements, platform DNS-SD state changes and failures.

Thanks @abtink ! It's really important but missing in OTBR!

File: include/openthread/platform/dnssd.h
Comment (wgtdkp) [inline include/openthread/platform/dnssd.h:102]: I think `mServiceType` should be more accurate. "Service Name" standards for only "_mt" in the example

Comment (wgtdkp) [reply include/openthread/platform/dnssd.h:102]: This also aligns with the name "SubType"

File: include/openthread/platform/dnssd.h
Comment (wgtdkp) [inline include/openthread/platform/dnssd.h:201]: Great documentation! :+1:

File: include/openthread/platform/dnssd.h
Comment (wgtdkp) [inline include/openthread/platform/dnssd.h:151]: Should we pass in the InfraIf index to make it clear that this service MUST be published on the infrastructure network?

Another reason is that we want to support changing infra if at runtime on Android, and it would be simpler to support this case if the Infra If index is provided here.

File: include/openthread/platform/dnssd.h
Comment (abtink) [reply include/openthread/platform/dnssd.h]: Good suggestion. Thanks. I renamed it to `mServiceType` in the new push.

File: include/openthread/platform/dnssd.h
Comment (abtink) [reply include/openthread/platform/dnssd.h:151]: I think it may be better to not include the infra-if index in the DNSSD platform APIs.

If we include the infra-if index in the `otPlatDnssd` APIs, we need the platform to somehow tell us which infra-if index to use. Within the OT core itself, we don't really need to know the infra-if index (other than passing it back as a parameter in platform APIs).

It may be simpler to assume that the platform should know the infra-netif (or maybe even a set of netifs) that it should do mDNS/DNS-SD on.

To support the use case where the infra-if changes at runtime with the DNS-SD platform APIs, we can signal such a change from the platform as a "state change" using `otPlatDnssdStateHandleStateChange()`. This should then trigger the OT stack to re-register all mDNS/DNS-SD entries (services/hosts) on new infra-if. 

One more related thought on this:
- In the `otPlatInfraIf` APIs, we do have the infra-If index as input, but this is tied to the BR and specifically `RoutingManager` functionality.
- The way we learn the infra-if index is actually through a public BR API `otBorderRoutingInit()`
- If we try to use the same infra-if index for DNSSD platform API, we indirectly restrict the use of the `otPlatDnssd` APIs to the use of the `RoutingManager` (which may not be good).

Thoughts?

**Comment by abtink** [comment]
The new commit makes large changes to the implementation:

- It adds new `otPlatDnssd` APIs to register and unregister key records.
- These APIs are used by `AdvertisingProxy` to keep the claim on the name of a removed host or service while its key lease is not expired.
- They are also used when an SRP host registration has no off-mesh routable address (when the host and its services are un-advertised).
- The unit test is updated to validate the key registration behavior.
- A new detailed test case is added that covers host changing its addresses (adding/removing off-mesh routable address).
- Another new test case is added that covers corner cases where an outstanding advertisement is removed before it is committed.
- Common code patterns used in `AdvertisingProxy` are now defined as template <Entry> methods, which are then used for `Host` or `Service` types (simplifying the code and removing repeated patterns).

File: include/openthread/platform/dnssd.h
Comment (wgtdkp) [reply include/openthread/platform/dnssd.h:167]: > - In the otPlatInfraIf APIs, we do have the infra-If index as input, but this is tied to the BR and specifically RoutingManager functionality.
> - The way we learn the infra-if index is actually through a public BR API otBorderRoutingInit()
> - If we try to use the same infra-if index for DNSSD platform API, we indirectly restrict the use of the otPlatDnssd APIs to the use of the RoutingManager (which may not be good).

The Infra interface passed to otPlatDnssd APIs doesn't necessary to be always the AIL selected for border routing. The otPlatDnssd APIs can be reused for TREL which can be functioning on another infra interface. For example, RoutingManager works on wlan0 but TREL works on eth0. This is allowed, right?

For border routing, it's better to advertise SRP service on an AIL only when we can provide connectivity between Thread and this AIL. If we don't pass a Infra interface parameter with the otPlatDnssd APIs, the implementations may just advertise the services on all AILs available on this host (OTBR today is doing so) and we certainly want to help avoid this by telling the implementer which exact infra interface he/she should use. Thoughts?

Comment (abtink) [reply include/openthread/platform/dnssd.h:167]: Thinking more on this, I agree it may be useful to sepcify the infra if index. 
Will add it as `mInfraIfIndex` variable in the `otPlatDnssdHost`, `Servcie` or `Key` structs (instead of an API input).

Comment (wgtdkp) [reply include/openthread/platform/dnssd.h:167]: Thanks @abtink !

> Will add it as mInfraIfIndex variable in the otPlatDnssdHost, Servcie or Key structs (instead of an API input).

This works for me, though I don't think `mInfraIfIndex` look like a property of a Host or Service - for example, if we change the advertising network interface, we shouldn't need to make update to the Host or Service objects.

Comment (abtink) [reply include/openthread/platform/dnssd.h:167]: My plan is to reuse the `otPlatDnssdService` and `otPlatDnssdHost` definitions in the "subscribe/browse" DNS-SD APIs (will add later). for use by discovery proxy, etc. The `mInfraIfIndex` property of these structs can be helpful for these APIs as well, and it can be used to learn which interface the entry was discovered on. We have something similar in the `otbr::Mdns::Publisher::DiscoveredInstanceInfo` struct.

Also, these types are used as temporary stack objects that are constructed to pass to the API. We do not retain them afterwards.

Comment (abtink) [reply include/openthread/platform/dnssd.h:170]: Added this in the new pushed commit.

**Comment by abtink** [comment]
In the new push, the implementation was updated to change how registrations are handled when the SRP client only includes mesh-local/link-local addresses. In the previous model, such services were not advertised on the AIL. In the new model, services are still advertised on the AIL, but without any addresses. This means that the service (SRV/TXT record) can be seen and queried on the AIL, but there will be no AAAA records for the hostname. This change is related to discussions in [SPEC-1211](https://threadgroup.atlassian.net/browse/SPEC-1211). For now, I kept the new changes as a separate commit in this PR (will later squash them into one).

File: doc/ot_api_doc.h
Comment (jwhui) [inline doc/ot_api_doc.h:187]: nit: make alphabetical?

File: src/core/net/srp_advertising_proxy.hpp
Comment (jwhui) [inline src/core/net/srp_advertising_proxy.hpp:122]:
```suggestion
     * @param[in] aHost      The `aHost` instance constructed from processing a newly received SRP Update message.
     * @param[in] aMetadata  The `MessageMetadata` associated with the received SRP Update message by server.
```

File: src/core/net/dns_types.hpp
Comment (jwhui) [inline src/core/net/dns_types.hpp:2077]:
```suggestion
     * Returns the ECDSA P-256 public key.
```

File: src/core/net/dns_types.hpp
Comment (jwhui) [inline src/core/net/dns_types.hpp:2087]:
```suggestion
     * Sets the ECDSA P-256 public key.
     *
     * @param[in] aKey  The public key.
```

File: src/core/net/srp_advertising_proxy.cpp
Comment (jwhui) [inline src/core/net/srp_advertising_proxy.cpp:762]:
```suggestion
        // `aExistingHost` we replace it with the request ID of the
```

File: include/openthread/platform/dnssd.h
Comment (abtink) [reply include/openthread/platform/dnssd.h:104]: Updated in new push to use `Length`.

File: src/core/net/srp_advertising_proxy.cpp
Comment (abtink) [reply src/core/net/srp_advertising_proxy.cpp:71]: Not sure. Adv-proxy goes hand-in-hand with SRP server and SRP server uses `LogInfo`.

File: include/openthread/platform/dnssd.h
Comment (abtink) [reply include/openthread/platform/dnssd.h:208]: This is a guideline to help platform implementations, in that they don't need to actively check for duplicate registrations (with not change at all). The platform implementation can still do check for duplicate if they want to but can assume that whenever this API is called something has changed. Thus making it easier on platform implementors.

We can easily ensure this in the OT stack, as we already do in the adv-proxy implementation from this PR. We track the registered SRP host and services, as well as which ones are already advertised. If there is no change, there is no need to call the API to register again. The same would apply for other use-cases in future (TREL or border agent).

File: include/openthread/platform/dnssd.h
Comment (abtink) [reply include/openthread/platform/dnssd.h]: Technically yes, but not sure if we have such a use-case today.

I updated the text to make it clear. The main point here is that platform implementation MUST not filter or remove any addresses. Any addresses that need to be filtered (e.g. link-local/mesh-local addresses of an registered SRP host) would be done by the OT stack itself before calling the API.

File: src/core/net/dnssd.hpp
Comment (abtink) [reply src/core/net/dnssd.hpp:102]: Exactly.

It represents a range of "Request ID" (from a start ID value up to to an end value).

- It is relatively small data structure so we can include it in every SRP `Host` entry (without adding too much size overhead). 
- We use it when we get a callback for a request ID to skip over some `Host` (and `Service` entries). If none of services of a `Host` is waiting for the ID, we can skip iterating through its service.
- And since the request IDs are allocated sequentially, the request IDs associated with same `Host` often fit in a short range.

File: src/core/net/srp_advertising_proxy.cpp
Comment (superwhd) [inline src/core/net/srp_advertising_proxy.cpp:1324]: IIUC the `mBlockingAdv` only blocks marking the completion of the advertisement, while the mDNS operations are not blocked. I wonder if it also needs to block the mDNS operations?

Comment (abtink) [reply src/core/net/srp_advertising_proxy.cpp:1324]: We don't need to block mDSN (`otPlatDnssd`) operations. The `CompareAndUpdateHostAndServices()` method already takes care of this. This method determines if the new entries (in a new SRP update) replace, change, or remove any existing entries (whether already registered or still being registered) and handles all cases.

The `mBlockingAdv` is used to ensure that "SRP Update" messages that are related (i.e., they change the same host or services) are committed in the order that they are received. When we set `mBlockingAdv` on an `AdvInfo`, we make sure it won't be committed even if its advertisement finishes early and before the earlier related one is committed.

Example situation:
- We receive an SRP Update from a client registering two services, A and B.
- While waiting for the this to advertised, we receive a new SRP Update that removes service B.
- Unregistering the service on `DnssdPlat` is often quick (no need probe), so the new advertisement for removing B will complete quickly.
- We need to make sure that these two SRP updates are committed in order.
- So we mark the second one to be blocked by the first one.
- Otherwise, we could end up with a situation where a removed service (B in this example) is incorrectly committed on the SRP server.

This situation of getting back-to-back messages that modify the same host/services will become a more common scenario with SRP Replication (when a new SRPL partner syncs, it gets all messages from the peer from the same client).

Comment (abtink) [reply src/core/net/srp_advertising_proxy.cpp:1324]: BTW, the new unit test `test_srp_adv_proxy` added in this PR, covers a bunch of these scenarios in `TestSrpAdvProxyReplacedEntries()`.

**Review by superwhd — APPROVED** [review]
Looks great, thanks!

**Comment by wgtdkp** [comment]
Hi @abtink , what's your plan on this PR as well as https://github.com/openthread/ot-br-posix/pull/2042? I suppose you still want to check in the two changes, but both are conflicted now.

**Comment by abtink** [comment]
> Hi @abtink , what's your plan on this PR as well as https://github.com/openthread/ot-br-posix/pull/2042? I suppose you still want to check in the two changes, but both are conflicted now.

I'd like to wait till we have the new PR on mDNS in OT core up and then update this to work with new APIs defined by it.
We keep the option of platform providing mDNS but want to make sure the API definitions are aligned.

**Comment by abtink** [comment]
Rebased on `main`. We can move with final review and merge of this.

**Review by wgtdkp — APPROVED** [review]
Thanks abtink 👍 
This is really a great work!
Especially the comprehensive tests in tests/unit/test_srp_adv_proxy.cpp are what missing in ot-br-posix. I think we will deprecate the impl in ot-br-posix once this is merged.

File: src/core/net/srp_advertising_proxy.cpp
Comment (superwhd) [inline src/core/net/srp_advertising_proxy.cpp:1108]: I just read [RFC3445 Section 3](https://www.rfc-editor.org/rfc/rfc3445.html#section-3) and found it's violating the following statement.
```
In the flags field, all bits except bit 7 are reserved and MUST be
   zero.  If Bit 7 (Zone bit) is set to 1, then the KEY is a DNS Zone
   key.  If Bit 7 is set to 0, the KEY is not a zone key.  SIG(0)/TKEY
   are examples of DNSSEC keys that are not zone keys.
```
Here `kSignatoryFlagGeneral` is 1 at the 16th bit.

I wonder if there's a following RFC that allows setting `1`s at bits other than the 7th bit?

Comment (abtink) [reply src/core/net/srp_advertising_proxy.cpp:1108]: This is following the same flags and format as in key RR used by SRP (e.g. in SRP client or checked in SRP server). Here we are just re-constructing the key RR that SRP sever got from client.

Regarding registering key on mDNS in adv-proxy, its purpose is to keep the claim to them name the record data is not really important. There were discussions of even allowing empty record data for it.

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:79]:
```suggestion
 * Represents the callback function used when registering/unregistering a host or service.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:93]:
```suggestion
 * Represents a DNS-SD service.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:115]:
```suggestion
 * Represents a DNS-SD host.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:130]:
```suggestion
 * Represents a DNS-SD key record.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:147]:
```suggestion
 * Callback from the platform to notify OpenThread of state changes to the DNS-SD module.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:149]:
```suggestion
 * The OpenThread stack will call `otPlatDnssdGetState()` (from this callback or later) to get the new state. The platform
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:158]:
```suggestion
 * Gets the current state of the DNS-SD module.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:160]:
```suggestion
 * The platform MUST notify OpenThread whenever its state gets changed by invoking `otPlatDnssdStateHandleStateChange()`.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:164]:
```suggestion
 * @returns The current state of the DNS-SD module.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:187]:
```suggestion
 * - The `mTxtData` and `mTxtDataLength` fields specify the encoded TXT data.
 * - The `mPort`, `mWeight`, and `mPriority` fields specify the service's parameters (as specified in DNS SRV record).
 * - The `mTtl` field specifies the TTL if non-zero. If zero, the platform can choose the TTL to use.
 * - The `mInfraIfIndex` field, if non-zero, specifies the infrastructure network interface index to use for this request. If
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:198]:
```suggestion
 * function. The OpenThread stack will ensure to handle such a situation.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:208]:
```suggestion
 * The OpenThread stack will not register the same service (with no changes) that was registered successfully earlier.
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:213]:
```suggestion
 * The OpenThread stack does not require the platform implementation to always invoke the @p aCallback function. The OpenThread
```

File: include/openthread/platform/dnssd.h
Comment (Vyrastas) [inline include/openthread/platform/dnssd.h:170]: All the suggestions in this comments for this method, can you also apply the same things to the rest of the methods here?  Mainly

OpenThread stack --> The OpenThread stack

and indicating "field" in each field bullet

**Review by Vyrastas — CHANGES_REQUESTED** [review]
Copyedit nits for the API, everything else looks good

File: include/openthread/platform/dnssd.h
Comment (abtink) [reply include/openthread/platform/dnssd.h:58]: Thanks I changed it to "APIs are used". 

Since it is a platform API, we expect platform to provide it and we (OT stack) will use it and we want to document under which build config these APIs are used by OT stack.

File: include/openthread/platform/dnssd.h
Comment (abtink) [reply include/openthread/platform/dnssd.h:170]: Thanks. 👍. Made all the changes.

**Review by Vyrastas — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**
